lib/jxl/memory_manager_internal.cc

```diff
@@ -113,15 +114,22 @@ AlignedMemory::AlignedMemory(JxlMemoryManager* memory_manager, void* allocation)
   size_t group =
       static_cast<size_t>(next_group.fetch_add(1, std::memory_order_relaxed));
   group &= (memory_manager_internal::kNumAlignmentGroups - 1);
-  size_t offset = memory_manager_internal::kAlignment * group;
+
+  size_t pre_padding_groups = pre_padding / memory_manager_internal::kAlignment;
+  size_t pre_padding_offset =
+      pre_padding_groups * memory_manager_internal::kAlignment;
+
+  size_t offset =
+      memory_manager_internal::kAlignment * group + pre_padding_offset;
 
   // Actual allocation.
   uintptr_t address = reinterpret_cast<uintptr_t>(allocation);
```

Comment:
Wouldn't just adding pre_padding to address here and doing everything else the same be enough? Why do we need to manipulate offset at all?

Reply:
Specifically adding pre_padding might cause the resulting address to not be aligned, which is not great...

This way, we get at most one extra kAlignment block of additional padding. This version of the code also works if pre_padding is larger than kAlias (just changing the check at line 132 would work otherwise).